utils/s2n_mem.c

```diff
@@ -25,23 +25,106 @@
 #include "utils/s2n_safety.h"
 
 static long page_size = 4096;
-static int use_mlock = 1;
 
-int s2n_mem_init(void)
+static int s2n_mem_init_impl(void);
+static int s2n_mem_cleanup_impl(void);
+static int s2n_mem_free_no_mlock_impl(void *ptr, uint32_t size);
+static int s2n_mem_free_mlock_impl(void *ptr, uint32_t size);
+static int s2n_mem_malloc_no_mlock_impl(void **ptr, uint32_t requested, uint32_t *allocated);
+static int s2n_mem_malloc_mlock_impl(void **ptr, uint32_t requested, uint32_t *allocated);
+
+static s2n_mem_init_callback s2n_mem_init_cb = s2n_mem_init_impl;
+static s2n_mem_cleanup_callback s2n_mem_cleanup_cb = s2n_mem_cleanup_impl;
+static s2n_mem_malloc_callback s2n_mem_malloc_cb = s2n_mem_malloc_mlock_impl;
+static s2n_mem_free_callback s2n_mem_free_cb = s2n_mem_free_mlock_impl;
+
+static int s2n_mem_init_impl(void)
 {
     GUARD(page_size = sysconf(_SC_PAGESIZE));
     if (getenv("S2N_DONT_MLOCK")) {
-        use_mlock = 0;
+        s2n_mem_malloc_cb = s2n_mem_malloc_no_mlock_impl;
+        s2n_mem_free_cb = s2n_mem_free_no_mlock_impl;
     }
-
-    return 0;
+    return S2N_SUCCESS;
 }
 
-int s2n_mem_cleanup(void)
+static int s2n_mem_cleanup_impl(void)
 {
     page_size = 4096;
-    use_mlock = 1;
-    return 0;
+    s2n_mem_malloc_cb = s2n_mem_malloc_no_mlock_impl;
+    s2n_mem_free_cb = s2n_mem_free_no_mlock_impl;
+    return S2N_SUCCESS;
+}
+
+static int s2n_mem_free_mlock_impl(void *ptr, uint32_t size)
+{
+    int munlock_rc = munlock(ptr, size);
+    free(ptr);
+    GUARD(munlock_rc);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_mem_free_no_mlock_impl(void *ptr, uint32_t size)
+{
+    free(ptr);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_mem_malloc_mlock_impl(void **ptr, uint32_t requested, uint32_t *allocated)
```

Comment:
This doesn't zero the memory.  Is that a problem, or intended?

Reply:
It's intended. After staring at a bunch of traces I finally figured out that zeroing memory on initialization isn't our concern; that's not s2n data that we ever had to protect. So we zero on release. I amended the unit test to initialize memory with all 'a' values, just to prove that nothing is silently depending on zero initialization.

Reply:
The advantage would be to make the code more deterministic, since right now something that invalidly reads new memory can read anything. But this might not be worth the cost of the memset.

Reply:
I'm definitely keen to avoid the memset, because I really want to keep the memset at the other end, and having 2 seems like overkill.